### PR TITLE
GET_ORGANIZATION permission for Env teams

### PR DESCRIPTION
**Pull request by SofiaSazonova**
### Feature or Bugfix
- Bugfix

### Detail
-  When Environment is linked, check that SamlGroup is invited to Organization
- When group is invited to Env, add GET_ORGANIZATION permissions, if it doesn't exist yet
- When group is deleted from Env Teams, check 1)it's not invited to Org, 2)it's not invited to another Env of this Org -> remove GET_ORGANIZATION permissions
- Related tests
- list_organisations now is based on GET_ORGANIZATION permissions, so non-mebers with GET_ORGANIZATION permissios can see Orgs 
- Inside OrgView all buttons (link, Invite, update, delete) are hidden basing on permissions
- migration script for permission backfill

### Relates
- #1147

### Security
Please answer the questions below briefly where applicable, or write `N/A`. Based on
[OWASP 10](https://owasp.org/Top10/en/).

- Does this PR introduce or modify any input fields or queries - this includes
fetching data from storage outside the application (e.g. a database, an S3 bucket)?
  - Is the input sanitized?
  - What precautions are you taking before deserializing the data you consume?
  - Is injection prevented by parametrizing queries?
  - Have you ensured no `eval` or similar functions are used?
- Does this PR introduce any functionality or component that requires authorization?
  - How have you ensured it respects the existing AuthN/AuthZ mechanisms?
  - Are you logging failed auth attempts?
- Are you using or adding any cryptographic features?
  - Do you use a standard proven implementations?
  - Are the used keys controlled by the customer? Where are they stored?
- Are you introducing any new policies/roles/users?
  - Have you used the least-privilege principle? How?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by SofiaSazonova** [comment]
Checked in AWS:
Org, Env1, Env2, Group (not in Org)

- [ ] Can't create an Env with Group
- [ ] Can create Env with group that is invited into Org

- [ ] When Group is invited to the Env1, it gets GET_ORGANIZATION permissions for Org
- [ ] When Group is invited to the Env2, GET_ORGANIZATION permission for Org is not duplicated

- [ ] When Group is deleted from Env2,  GET_ORGANIZATION permission for Org is NOT deleted
- [ ] When Group is deleted from Env1,  GET_ORGANIZATION permission for Org is deleted

- [ ] Group is invited into Org, then invited into Env1, GET_ORGANIZATION permission for Org is not duplicated
- [ ] When Group is deleted from Env1,  GET_ORGANIZATION permission for Org is NOT deleted
- [ ] When Group is deleted from Org,  GET_ORGANIZATION permission for Org is deleted

File: backend/dataall/core/environment/services/environment_service.py [hunk 1]
```diff
@@ -213,6 +215,12 @@ def create_environment(uri, data=None):
         context = get_context()
         with context.db_engine.scoped_session() as session:
             EnvironmentRequestValidationService.validate_creation_params(data, uri, session)
+            if not OrganizationRepository.is_group_invited(session, uri, data['SamlGroupName']):
+                raise Exception(
+                    f'Group {data["SamlGroupName"]} is not a member of the organization {data.get("organizationUri")}. '
+                    f'Invite this group to the organisation before linking the environment.'
+                )
+
```
Comment (noah-paige) [inline backend/dataall/core/environment/services/environment_service.py:223]: Can we move to to part of `validate_user_groups()` in `EnvironmentRequestValidationService`

File: backend/dataall/core/organizations/db/organization_repositories.py [hunk 2]
```diff
@@ -106,6 +106,19 @@ def query_organization_groups(session, uri, filter) -> Query:
...
+        return (
+            session.query(models.OrganizationGroup)
+            .filter(
+                and_(
+                    models.OrganizationGroup.organizationUri == uri,
+                    models.OrganizationGroup.groupUri == group,
+                )
+            )
+            .count()
+        ) > 0
```
Comment (noah-paige) [inline backend/dataall/core/organizations/db/organization_repositories.py:120]: nit: Don't need to specify `>0` - if returns a value greather than 0 will evaluate as True in the if condition either way

File: backend/dataall/core/environment/db/environment_repositories.py [hunk 3]
```diff
@@ -303,6 +303,22 @@ def is_user_invited_to_environment(session, groups, uri):
...
+            .join(Environment, EnvironmentGroup.environmentUri == Environment.environmentUri)
+            .filter(Environment.deleted.is_(None))
+            .filter(
+                and_(
+                    EnvironmentGroup.groupUri == group,
+                    Environment.organizationUri == organizationUri,
+                )
+            )
+            .count()
+            > 0
```
Comment (noah-paige) [inline backend/dataall/core/environment/db/environment_repositories.py:319]: nit: Don't need to specify >0 - if returns a value greather than 0 will evaluate as True in the if condition either way

File: backend/dataall/core/environment/services/environment_service.py [hunk 4]
```diff
@@ -463,6 +487,20 @@ def remove_group(uri, group):
                 resource_uri=environment.environmentUri,
                 resource_type=Environment.__name__,
             )
+
+            # if group is no longer invited to organization or environments inside it, remove GET_ORGAISATION policy
```
Comment (noah-paige) [inline backend/dataall/core/environment/services/environment_service.py:491]: nit: type in comment `remove GET_ORGANIZATION policy`

File: backend/dataall/core/environment/services/environment_service.py [hunk 5]
```diff
@@ -409,6 +417,22 @@ def invite_group(uri, data=None) -> (Environment, EnvironmentGroup):
...
+                resource_type=Organization.__name__,
+                permission_name=GET_ORGANIZATION,
+            ):
+                ResourcePolicyService.attach_resource_policy(
+                    session=session,
+                    group=group,
+                    resource_uri=environment.organizationUri,
+                    permissions=[GET_ORGANIZATION],
+                    resource_type=Organization.__name__,
+                )
```
Comment (noah-paige) [inline backend/dataall/core/environment/services/environment_service.py:434]: The way `def attach_resource_policy()` is written is that the policy is only created and permissions associated if and only if they do not already exist (like a "create if not exists" expressions)

Open ended Q - is there added benefit to having the check `if check_group_resource_permission()`?

**Comment by noah-paige** [comment]
Also a note - I believe there is a bug with the confirmation pop ups when removing an invited team or consumption role from an environment

Let's say I have 3 groups in an env (GroupA, B, and C)... When removing GroupA - I receive the pop up and need to type `permanently delete` for the group
- Once I do so GroupA is removed but the pop up remains and the team it is asking me to confirm to remove now is GroupB 

In `EnvironmentTeams.js` we have one variable `isDeleteTeamModalOpen` to control if the mondals for N groups is open - some bug in the logic is being introduced because of this

File: backend/dataall/core/environment/services/environment_service.py (hunk 5 above)
Comment (dlpzx) [reply backend/dataall/core/environment/services/environment_service.py:434]: True, attach_resource_policy does not create duplicates if the policy already exists. I don't think we need the check at creation, only at deletion

File: backend/dataall/core/environment/services/environment_service.py (hunk 1 above)
Comment (SofiaSazonova) [reply backend/dataall/core/environment/services/environment_service.py]: done

File: backend/dataall/core/environment/services/environment_service.py (hunk 4 above)
Comment (SofiaSazonova) [reply backend/dataall/core/environment/services/environment_service.py]: done

File: backend/dataall/core/environment/services/environment_service.py (hunk 5 above)
Comment (SofiaSazonova) [reply backend/dataall/core/environment/services/environment_service.py]: done

**Comment by SofiaSazonova** [comment]
@noah-paige 
(1) I think, it's for separate PR
(2) Done. I didn't remove REMOVE_ORGANIZATION_GROUP yet, but backfilled GET_ORGANISATION
(3) I changed OrganisationView. Now readers are displayed. Also, I changed list_organisation logic to show Orgs for "readers"
(4) I will fix this bug in separate PR

**Comment by SofiaSazonova** [comment]
image

File: backend/dataall/core/groups/services/group_service.py [hunk 6]
```diff
@@ -72,6 +72,6 @@ def list_groups_without_invited(filter: dict = None):
             groups = []
             for group in cognito_groups:
                 if group not in invited_groups:
-                    groups.append({'groupName': group})
+                    groups.append({'groupUri': group})
```
Comment (noah-paige) [inline backend/dataall/core/groups/services/group_service.py:75]: why this change?

File: backend/dataall/core/groups/api/types.py [hunk 7]
```diff
@@ -45,6 +45,18 @@
...
+
+OrganizationReadersSearchResult = gql.ObjectType(
+    name='OrganizationReaders',
+    fields=[
+        gql.Field(name='count', type=gql.Integer),
+        gql.Field(name='page', type=gql.Integer),
+        gql.Field(name='pages', type=gql.Integer),
+        gql.Field(name='hasNext', type=gql.Boolean),
+        gql.Field(name='hasPrevious', type=gql.Boolean),
+        gql.Field(name='nodes', type=gql.ArrayType(GroupsInfo)),
```
Comment (noah-paige) [inline backend/dataall/core/groups/api/types.py:60]: Is this type used anywhere?

File: backend/dataall/core/groups/api/types.py [hunk 8]
```diff
@@ -45,6 +45,18 @@
 GroupsInfo = gql.ObjectType(
     name='GroupsInfo',
     fields=[
-        gql.Field(name='groupName', type=gql.String),
+        gql.Field(name='groupUri', type=gql.String),
```
Comment (noah-paige) [inline backend/dataall/core/groups/api/types.py:48]: why change this from groupName to groupUri?

File: backend/dataall/core/organizations/db/organization_repositories.py [hunk 9]
```diff
@@ -106,6 +122,52 @@ def query_organization_groups(session, uri, filter) -> Query:
...
+            .join(
+                ResourcePolicyPermission,
+                ResourcePolicy.sid == ResourcePolicyPermission.sid,
+            )
+            .join(
+                Permission,
+                Permission.permissionUri == ResourcePolicyPermission.permissionUri,
+            )
+            .filter(
+                or_(
```
Comment (noah-paige) [inline backend/dataall/core/organizations/db/organization_repositories.py:140]: don't need the `or_(...)` condition

File: frontend/src/modules/Organizations/components/OrganizationTeamInviteForm.js [hunk 10]
```diff
@@ -48,8 +48,8 @@ export const OrganizationTeamInviteForm = (props) => {
         setGroupOptions(
           response.data.listGroups.map((g) => ({
             ...g,
-            value: g.groupName,
-            label: g.groupName
+            value: g.groupUri,
```
Comment (noah-paige) [inline frontend/src/modules/Organizations/components/OrganizationTeamInviteForm.js:51]: similar to above - any need to change `groupName` to `groupUri`?

File: backend/migrations/versions/11b716ddb036_backfill_get_org_permissions.py [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+from sqlalchemy import orm
+from dataall.core.organizations.db.organization_models import Organization
+from dataall.core.environment.db.environment_models import EnvironmentGroup, Environment
+from dataall.core.permissions.services.resource_policy_service import ResourcePolicyService
+from dataall.core.permissions.services.organization_permissions import GET_ORGANIZATION
+
+
+# revision identifiers, used by Alembic.
+revision = '11b716ddb036'
+down_revision = '458572580709'
```
Comment (noah-paige) [inline backend/migrations/versions/11b716ddb036_backfill_get_org_permissions.py:19]: I see the migration script is failing because of multiple heads in alembic ... I believe you will need to merge the latest from `main` to this branch and then update the `down_revision` here to be pointing from the latest revision currently in main

**Comment by noah-paige** [comment]
The new frontend UI looks great! 1 general question - does the query still run in manageable time for either `getOrganization` or `listOrganizationReadOnlyGroups` with the nested resolvers / added logic?

I was considering the idea that we could add a new column to `OrganizationGroup` called `userType` and backfiling for existing Groups. Then for each invite or remove from envs we also create/remove a new OrganizationGroup record with `userType` of value `Reader` 
- Could make it easier to determine what should be loaded for a given user on UI / when listing reader vs admin users 

... but this is not really needed unless we see some big performance differences I would argue

**Comment by SofiaSazonova** [comment]
Well, right now I can't spot any big difference. We have queries like this in many places. But of course, it's better to test with some loaded infrastructure.

As for OrganisationGroups. Well, we can do this, but it means, that we also need to redo the UI and backend of inviting groups: we want then to add opportunity to give Readonly permissions. 
May be, it's a good thing to do. I'm not sure right now. Let's discuss

**Comment by noah-paige** [comment]
Last required change I think - the resolved user role for an invite EnvGroup (i.e. an Org Reader but not an Administrator) shows as `NotMember`

Screenshot 2024-05-24 at 10 59 34 AM

Would need to update the way `find_organization_membership` identifies who is part of org group or not

File: backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py [hunk 12]
```diff
@@ -108,9 +108,9 @@ def find_resource_policy(session, group_uri: str, resource_uri: str, resource_ty
 
     @staticmethod
     def find_all_resource_policies(
-        session, group_uri: str, resource_uri: str, resource_type: str = None, permissions: List[str] = None
+        session, groups: [str], resource_uri: str, resource_type: str = None, permissions: List[str] = None
```
Comment (noah-paige) [inline backend/dataall/core/permissions/db/resource_policy/resource_policy_repositories.py:111]: nit: `List[str]`

**Review by noah-paige — APPROVED** [review]
Left 2 minor comments but overall tested and looks good
